### PR TITLE
[MRV2][Feature][Bugfix] Support and fix Top-K/Top-P Triton kernel on Ascend

### PR DESCRIPTION
**Pull request by AuroraEmiya**
### What this PR does / why we need it?

This PR adapts the fused Top-K/Top-P Triton kernel for Ascend NPU from upstream vllm and fixes several numerical and boundary correctness issues exposed during large-vocabulary testing.

The changes can be grouped into three categories:

#### 1. Functional fixes

* Adapt Triton reduction and prefix-scan operations to Ascend-supported integer types.
* Avoid unsupported scalar table-index lowering by using vectorized lookup for the Top-K/Top-P approximation tables.
* Use Ascend-specific tile sizes to avoid UB overflow when processing large vocabularies.

#### 2. Precision fixes

* Improve Top-P boundary handling against FP32 reduction and rounding effects.
* Prevent boundary-token count calculation from dropping the whole boundary group because of numerical rounding.

#### 3. Correctness fixes

* Explicitly handle duplicate values at Top-K/Top-P boundaries instead of relying on the strict pivot comparison to implicitly preserve them.
* Fix an inconsistent Top-P forced-termination state where the final pivot and its boundary statistics could come from different pivot evaluations. This could keep a lower-logit token while masking a higher-logit token near the Top-P boundary.

The main Top-K/Top-P algorithm and overall fused-kernel structure are preserved.

### Does this PR introduce *any* user-facing change?

No.

The external API and sampling semantics remain unchanged. This PR improves Ascend compatibility, numerical robustness, and correctness of the existing Top-K/Top-P implementation.

### How was this patch tested?

The Triton Top-K/Top-P UT covers:

* Top-K only
* Top-P only
* Combined Top-K + Top-P
* Large batch sizes and vocabularies
* Disabled Top-K / Top-P
* Non-contiguous logits
* `-inf` / grammar-masked logits
* Few-valid-token and equal-logit boundary cases
* Input contract and edge cases

The full Top-K/Top-P Triton UT passes successfully:

```bash
pytest -sv  \ #--noconftest
tests/e2e/nightly/single_node/ops/singlecard_ops/triton/test_apply_top_k_top_p_triton.py
```

- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/58d3918e3ea0a544ffedadad2ba84559e9c51d8f

### CODE REVIEW

**Review by ZT-AIA — DISMISSED** [review]

**Comment by JC-ut0** [comment]
## PR #13847 Review Feedback: A3 NPU Benchmark Results

### Test Environment

- A3 16-card NPU, torch 2.10.0, triton-ascend 3.2.0
- PR code loaded from GitHub via `sys.path`, confirmed importing PR's `apply_top_k_top_p_triton` (with NPU BLOCK_SIZE=1024 logic)
- 4-way comparison: Triton (PR), UpPT (upstream sort+scatter), VA-PT (vllm-ascend sort+mask), NPU (`npu_top_k_top_p`)
- 5 warmup iterations, 20 timed iterations, median reported

### Accuracy

| Scenario | Result |
|---|---|
| top-k only (all configs) | PASS, bit-exact |
| top-k + top-p | PASS |
| **top-p only (V=32000, p=0.9)** | **FAIL**, count_diff=5, exceeds tolerance of 3 |

The top-p standalone path has a boundary accuracy bug. Recommend investigating the forced-termination path's boundary statistics consistency.

### Performance (ms, median)

**Small k (1-100), no p:**

| B | V | K | Triton | UpPT | VA-PT | NPU | Best |
|---|---|---|---|---|---|---|---|
| 8 | 1024 | 1 | 0.214 | 0.239 | 0.194 | **0.086** | NPU |
| 8 | 32000 | 1 | 1.618 | 0.802 | 0.172 | **0.130** | NPU |
| 8 | 128256 | 1 | 5.999 | 2.641 | 0.341 | **0.266** | NPU |
| 32 | 128256 | 1 | 11.868 | 13.662 | 0.589 | **0.466** | NPU |

**top-p only:**

| B | V | P | Triton | UpPT | VA-PT | NPU | Best |
|---|---|---|---|---|---|---|---|
| 8 | 32000 | 0.9 | 2.850 | 0.846 | **0.293** | 1.104 | VA-PT |
| 8 | 32000 | 0.5 | 2.528 | 0.847 | **0.295** | 1.079 | VA-PT |
| 8 | 128256 | 0.9 | 10.911 | 2.905 | **0.618** | 4.294 | VA-PT |
| 32 | 128256 | 0.9 | 21.518 | 13.950 | **0.911** | 14.736 | VA-PT |

**k + p combined:**

| B | V | K | P | Triton | UpPT | VA-PT | NPU | Best |
|---|---|---|---|---|---|---|---|---|
| 8 | 128256 | 20 | 0.9 | 6.103 | 2.943 | 0.662 | **0.272** | NPU |
| 8 | 128256 | 20 | 0.5 | 6.096 | 2.947 | 0.660 | **0.270** | NPU |
| 32 | 128256 | 20 | 0.9 | 12.038 | 14.032 | 0.986 | **0.475** | NPU |

**Large k:**

| B | V | K | Triton | UpPT | VA-PT | NPU | Best |
|---|---|---|---|---|---|---|---|
| 8 | 128256 | 50000 | 14.117 | 2.629 | **0.332** | 1.774 | VA-PT |
| 8 | 128256 | 100000 | 23.590 | 2.628 | **0.333** | 3.182 | VA-PT |
| 32 | 128256 | 50000 | 27.507 | 13.660 | **0.594** | 9.035 | VA-PT |
| 128 | 128256 | 50000 | 83.877 | 41.272 | **1.975** | 27.278 | VA-PT |

**Mixed batch (half small k, half k=V):**

| B | V | K | P | Triton | UpPT | VA-PT | NPU | Best |
|---|---|---|---|---|---|---|---|---|
| 8 | 32000 | mix(1+V) | - | 1.625 | 0.798 | **0.170** | 1.154 | VA-PT |
| 8 | 128256 | mix(1+V) | - | 5.994 | 2.631 | **0.333** | 4.050 | VA-PT |
| 32 | 128256 | mix(1+V) | - | 6.019 | 13.660 | **0.585** | 21.846 | VA-PT |
| 8 | 128256 | mix(1+V) | 0.9 | 10.879 | 2.946 | **0.660** | 3.256 | VA-PT |
| 32 | 128256 | mix(1+V) | 0.9 | 16.959 | 14.023 | **0.984** | 14.334 | VA-PT |
| 32 | 128256 | mix(1+V) | 0.5 | 15.518 | 14.032 | **0.986** | 5.479 | VA-PT |

### Analysis

1. **The Triton kernel is not optimal in any of the 36 tested scenarios.** The primary cause is the NPU BLOCK_SIZE set to 1024, resulting in 126 tile iterations for V=128K — the loop overhead far exceeds the pivot algorithm's savings.

2. **Missing baselines.** The PR only compares against upstream `apply_top_k_top_p_pytorch` (sort+scatter). It should also benchmark:
   - `npu_top_k_top_p` (CANN native op, 0.08-0.47ms for small k)
   - `_apply_top_k_top_p_pytorch` (vllm-ascend sort+mask, 0.17-2ms for large k / top-p / mixed batch)

3. **`npu_top_k_top_p` is extremely fast for small k** (0.08-0.47ms), but degrades with large k (3ms at k=100K) and mixed batch (22ms at B=32 mix k=1+V).

4. **VA-PT (sort+mask) has stable latency** (0.17-2ms), independent of k/p values. It wins across top-p only, large k, and mixed batch scenarios.

5. **BLOCK_SIZE tuning recommendation.** The current `min(1024, max(128, next_power_of_2(vocab_size)))` is too conservative. Testing 2048/4096 on A3 could significantly reduce tile count (V=128K from 126 tiles down to 32/16).

6. **Accuracy bug.** top-p standalone (V=32000, p=0.9) has count_diff=5 exceeding tolerance. Recommend investigating the forced-termination path.

### Suggestions

- Retest performance after increasing BLOCK_SIZE (2048/4096)
- Fix the top-p accuracy bug
- Consider a hybrid strategy: small k via `npu_top_k_top_p`, other cases via sort+mask or Triton (after tuning)
- Add `npu_top_k_top_p` and `_apply_top_k_top_p_pytorch` as baselines in the PR description

**Comment by AuroraEmiya** [comment]
/rerun
[Bot]: rerun completed.

**Rerun:**
- E2E

**Comment by AuroraEmiya** [comment]
> ## PR #13847 Review Feedback: A3 NPU Benchmark Results
> ### Test Environment
> * A3 16-card NPU, torch 2.10.0, triton-ascend 3.2.0
> * PR code loaded from GitHub via `sys.path`, confirmed importing PR's `apply_top_k_top_p_triton` (with NPU BLOCK_SIZE=1024 logic)
> * 4-way comparison: Triton (PR), UpPT (upstream sort+scatter), VA-PT (vllm-ascend sort+mask), NPU (`npu_top_k_top_p`)
> * 5 warmup iterations, 20 timed iterations, median reported
> 
> ### Accuracy
> Scenario	Result
> top-k only (all configs)	PASS, bit-exact
> top-k + top-p	PASS
> **top-p only (V=32000, p=0.9)**	**FAIL**, count_diff=5, exceeds tolerance of 3
> The top-p standalone path has a boundary accuracy bug. Recommend investigating the forced-termination path's boundary statistics consistency.
> 
> ### Performance (ms, median)
> **Small k (1-100), no p:**
> 
> B	V	K	Triton	UpPT	VA-PT	NPU	Best
> 8	1024	1	0.214	0.239	0.194	**0.086**	NPU
> 8	32000	1	1.618	0.802	0.172	**0.130**	NPU
> 8	128256	1	5.999	2.641	0.341	**0.266**	NPU
> 32	128256	1	11.868	13.662	0.589	**0.466**	NPU
> **top-p only:**
> 
> B	V	P	Triton	UpPT	VA-PT	NPU	Best
> 8	32000	0.9	2.850	0.846	**0.293**	1.104	VA-PT
> 8	32000	0.5	2.528	0.847	**0.295**	1.079	VA-PT
> 8	128256	0.9	10.911	2.905	**0.618**	4.294	VA-PT
> 32	128256	0.9	21.518	13.950	**0.911**	14.736	VA-PT
> **k + p combined:**
> 
> B	V	K	P	Triton	UpPT	VA-PT	NPU	Best
> 8	128256	20	0.9	6.103	2.943	0.662	**0.272**	NPU
> 8	128256	20	0.5	6.096	2.947	0.660	**0.270**	NPU
> 32	128256	20	0.9	12.038	14.032	0.986	**0.475**	NPU
> **Large k:**
> 
> B	V	K	Triton	UpPT	VA-PT	NPU	Best
> 8	128256	50000	14.117	2.629	**0.332**	1.774	VA-PT
> 8	128256	100000	23.590	2.628	**0.333**	3.182	VA-PT
> 32	128256	50000	27.507	13.660	**0.594**	9.035	VA-PT
> 128	128256	50000	83.877	41.272	**1.975**	27.278	VA-PT
> **Mixed batch (half small k, half k=V):**
> 
> B	V	K	P	Triton	UpPT	VA-PT	NPU	Best
> 8	32000	mix(1+V)	-	1.625	0.798	**0.170**	1.154	VA-PT
> 8	128256	mix(1+V)	-	5.994	2.631	**0.333**	4.050	VA-PT
> 32	128256	mix(1+V)	-	6.019	13.660	**0.585**	21.846	VA-PT
> 8	128256	mix(1+V)	0.9	10.879	2.946	**0.660**	3.256	VA-PT
> 32	128256	mix(1+V)	0.9	16.959	14.023	**0.984**	14.334	VA-PT
> 32	128256	mix(1+V)	0.5	15.518	14.032	**0.986**	5.479	VA-PT
> ### Analysis
> 1. **The Triton kernel is not optimal in any of the 36 tested scenarios.** The primary cause is the NPU BLOCK_SIZE set to 1024, resulting in 126 tile iterations for V=128K — the loop overhead far exceeds the pivot algorithm's savings.
> 2. **Missing baselines.** The PR only compares against upstream `apply_top_k_top_p_pytorch` (sort+scatter). It should also benchmark:
>    
>    * `npu_top_k_top_p` (CANN native op, 0.08-0.47ms for small k)
>    * `_apply_top_k_top_p_pytorch` (vllm-ascend sort+mask, 0.17-2ms for large k / top-p / mixed batch)
> 3. **`npu_top_k_top_p` is extremely fast for small k** (0.08-0.47ms), but degrades with large k (3ms at k=100K) and mixed batch (22ms at B=32 mix k=1+V).
> 4. **VA-PT (sort+mask) has stable latency** (0.17-2ms), independent of k/p values. It wins across top-p only, large k, and mixed batch scenarios.
> 5. **BLOCK_SIZE tuning recommendation.** The current `min(1024, max(128, next_power_of_2(vocab_size)))` is too conservative. Testing 2048/4096 on A3 could significantly reduce tile count (V=128K from 126 tiles down to 32/16).
> 6. **Accuracy bug.** top-p standalone (V=32000, p=0.9) has count_diff=5 exceeding tolerance. Recommend investigating the forced-termination path.
> 
> ### Suggestions
> * Retest performance after increasing BLOCK_SIZE (2048/4096)
> * Fix the top-p accuracy bug
> * Consider a hybrid strategy: small k via `npu_top_k_top_p`, other cases via sort+mask or Triton (after tuning)
> * Add `npu_top_k_top_p` and `_apply_top_k_top_p_pytorch` as baselines in the PR description

Thanks for the detailed review and performance testing! :)

This PR mainly aims to keep the Ascend implementation aligned with the corresponding vLLM upstream V1 operator, so the scope of this PR is currently limited to functional enablement and correctness fixes.

For Top-P accuracy, the UT follows the upstream tolerance: the kept-count difference should be either <= 3 or within 0.5%. The correctness issues found during the Ascend adaptation, including the Top-P boundary handling issue, have also been fixed and covered by UTs.

Performance optimization is not the focus of this PR yet. The benchmark results and suggestions around BLOCK_SIZE tuning, native NPU operators, and hybrid dispatch are very useful, and we can investigate them separately as follow-up optimization work.

Thanks again for the thorough benchmark and suggestions!

**Comment by AuroraEmiya** [comment]
/rerun
[Bot]: rerun completed.

**Rerun:**
- E2E

**Comment by AuroraEmiya** [comment]
/rerun
[Bot]: rerun completed.

**Rerun:**
- E2E

**Comment by AuroraEmiya** [comment]
/rerun
[Bot]: rerun completed.

**Rerun:**
- E2E

**Comment by AuroraEmiya** [comment]
/rerun
[Bot]: rerun completed. No failed jobs found.

**Comment by AuroraEmiya** [comment]
/rerun
[Bot]: rerun completed. No failed jobs found.

**Review by ZT-AIA — APPROVED** [review]

**Review by chenlongxiao — APPROVED** [review]
